cpp/include/raft/comms/mpi_comms.hpp

```diff
@@ -0,0 +1,290 @@
+      _next_request_id(0) {
+    int mpi_is_initialized = 0;
+    MPI_TRY(MPI_Initialized(&mpi_is_initialized));
+    RAFT_EXPECTS(mpi_is_initialized, "ERROR: MPI is not initialized!");
+    MPI_TRY(MPI_Comm_size(_mpi_comm, &_size));
+    MPI_TRY(MPI_Comm_rank(_mpi_comm, &_rank));
+    //get NCCL unique ID at rank 0 and broadcast it to all others
+    ncclUniqueId id;
+    if (0 == _rank) NCCL_TRY(ncclGetUniqueId(&id));
+    MPI_TRY(MPI_Bcast((void*)&id, sizeof(id), MPI_BYTE, 0, _mpi_comm));
+
+    //initializing NCCL
+    NCCL_TRY(ncclCommInitRank(&_nccl_comm, _size, id, _rank));
+  }
+
+  virtual ~mpi_comms() {
+    //finalizing NCCL
+    NCCL_TRY_NO_THROW(ncclCommDestroy(_nccl_comm));
+    if (_owns_mpi_comm) {
+      MPI_TRY_NO_THROW(MPI_Comm_free(&_mpi_comm));
+    }
+  }
+
+  int get_size() const { return _size; }
+
+  int get_rank() const { return _rank; }
+
+  std::unique_ptr<comms_iface> comm_split(int color, int key) const {
+    MPI_Comm new_comm;
+    MPI_TRY(MPI_Comm_split(_mpi_comm, color, key, &new_comm));
+    return std::unique_ptr<comms_iface>(new mpi_comms(new_comm, true));
+  }
+
+  void barrier() const { MPI_TRY(MPI_Barrier(_mpi_comm)); }
+
+  void isend(const void* buf, size_t size, int dest, int tag,
+             request_t* request) const {
+    MPI_Request mpi_req;
+    request_t req_id;
+    if (_free_requests.empty()) {
+      req_id = _next_request_id++;
+    } else {
+      auto it = _free_requests.begin();
+      req_id = *it;
+      _free_requests.erase(it);
+    }
+    MPI_TRY(MPI_Isend(buf, size, MPI_BYTE, dest, tag, _mpi_comm, &mpi_req));
+    _requests_in_flight.insert(std::make_pair(req_id, mpi_req));
+    *request = req_id;
+  }
+
+  void irecv(void* buf, size_t size, int source, int tag,
+             request_t* request) const {
+    MPI_Request mpi_req;
+    request_t req_id;
+    if (_free_requests.empty()) {
+      req_id = _next_request_id++;
+    } else {
+      auto it = _free_requests.begin();
+      req_id = *it;
+      _free_requests.erase(it);
+    }
+
+    MPI_TRY(MPI_Irecv(buf, size, MPI_BYTE, source, tag, _mpi_comm, &mpi_req));
+    _requests_in_flight.insert(std::make_pair(req_id, mpi_req));
+    *request = req_id;
+  }
+
+  void waitall(int count, request_t array_of_requests[]) const {
+    std::vector<MPI_Request> requests;
+    requests.reserve(count);
+    for (int i = 0; i < count; ++i) {
+      auto req_it = _requests_in_flight.find(array_of_requests[i]);
+      RAFT_EXPECTS(_requests_in_flight.end() != req_it,
+                   "ERROR: waitall on invalid request: %d",
+                   array_of_requests[i]);
+      requests.push_back(req_it->second);
+      _free_requests.insert(req_it->first);
+      _requests_in_flight.erase(req_it);
+    }
+    MPI_TRY(MPI_Waitall(requests.size(), requests.data(), MPI_STATUSES_IGNORE));
+  }
+
+  void allreduce(const void* sendbuff, void* recvbuff, size_t count,
+                 datatype_t datatype, op_t op, cudaStream_t stream) const {
```

Comment:
This is just FYI,

Unified Collective Communication is coming (UCX is for P2P, UCC supports collectives on top of UCX).

https://www.ucfconsortium.org/projects/ucc/

At some point, we may consider implementing something like

```
collective(address) {
  if (address is device) {
    use NCCL;
  }
  else { // address is host
    // this is a temporary hack, we plan to replace the code below with UCC in the future.
    // communicator may own a small buffer. If the message size is smaller than the buffer size, we can use the buffer to avoid device memory allocation.
    create device_buffer;
    copy_to_device_buffer;
    use NCCL;
    copy back to host;
  }
}
```

I think this will make our code simpler and future ready if we need run collectives on host data.